File: backend/app/schemas/album.py
from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime
from typing import Optional, List
from .photo import PhotoResponse


class AlbumBase(BaseModel):
    title: str = Field(..., min_length=2, max_length=200, description="Название альбома")
    description: Optional[str] = Field(None, max_length=1000, description="Описание альбома")


class AlbumCreate(AlbumBase):
    family_id: int = Field(..., description="ID семьи")
    event_id: Optional[int] = Field(None, description="ID события (опционально)")


class AlbumUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=1000)


class AlbumMemberInfo(BaseModel):
    user_id: int
    can_edit: bool
    can_delete: bool
    status: str
    added_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AlbumResponse(AlbumBase):
    id: int
    family_id: int
    created_by_user_id: int
    event_id: Optional[int]
    created_at: datetime
    expires_at: datetime
    is_deleted: bool
    hours_until_deletion: float
    photos_count: int = 0
    members_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class AlbumDetailResponse(AlbumResponse):
    members: List[AlbumMemberInfo] = []
    photos: List[PhotoResponse] = []
    is_admin: bool = False


class AlbumListResponse(BaseModel):
    albums: List[AlbumResponse]
    total: int

File: frontend/src/api/album.ts
import { apiClient } from './client';

export interface Album {
  id: number;
  title: string;
  description: string;
  family_id: number;
  created_by_user_id: number;
  event_id: number | null;
  created_at: string;
  expires_at: string;
  is_deleted: boolean;
  hours_until_deletion: number;
  photos_count: number;
  members_count: number;
}

export interface CreateAlbumData {
  title: string;
  description?: string;
  family_id: number;
  event_id?: number;
}

export interface AlbumsResponse {
  albums: Album[];
  total: number;
}

export interface AlbumMember {
  album_id: number;
  user_id: number;
  can_edit: boolean;
  can_delete: boolean;
  status: string;
  added_by_user_id: number;
  added_at: string;
  user?: {
    id: number;
    name: string;
    email: string;
  };
}

export const albumAPI = {
  create: (data: CreateAlbumData) =>
    apiClient.post<Album>('/albums', data),

  getByFamily: (familyId: number) =>
    apiClient.get<AlbumsResponse>(`/albums?family_id=${familyId}`),

  getById: (albumId: number) =>
    apiClient.get<Album>(`/albums/${albumId}`),

  update: (albumId: number, data: { title?: string; description?: string }) =>
    apiClient.put<Album>(`/albums/${albumId}`, data),

  delete: (albumId: number) =>
    apiClient.delete(`/albums/${albumId}`),

  getViewers: (albumId: number) =>
    apiClient.get<AlbumMember[]>(`/albums/${albumId}/viewers`),

  getAdmins: (albumId: number) =>
    apiClient.get<AlbumMember[]>(`/albums/${albumId}/admins`),

  addMember: (albumId: number, userId: number) =>
    apiClient.post(`/albums/${albumId}/members`, { user_id: userId }),

  removeMember: (albumId: number, userId: number) =>
    apiClient.delete(`/albums/${albumId}/members/${userId}`),

  addAdmin: (albumId: number, userId: number) =>
    apiClient.post(`/albums/${albumId}/admins`, { user_id: userId }),

  removeAdmin: (albumId: number, userId: number) =>
    apiClient.delete(`/albums/${albumId}/admins/${userId}`),

  
};